DUI3-DX/Connectors/Rhino/Speckle.Connectors.Rhino7/Bindings/RhinoSendBinding.cs

```diff
@@ -0,0 +1,380 @@
+    catch (OperationCanceledException)
+    {
+      return;
+    }
+    catch (Exception e) when (!e.IsFatal()) // All exceptions should be handled here if possible, otherwise we enter "crashing the host app" territory.
+    {
+      _basicConnectorBinding.Commands.SetModelError(modelCardId, e);
+    }
+  }
+
+  public void CancelSend(string modelCardId) => CancellationManager.CancelOperation(modelCardId);
+
+  private Base ConvertObjects(
+    List<RhinoObject> rhinoObjects,
+    ISpeckleConverter converter,
+    SenderModelCard modelCard,
+    CancellationTokenSource cts
+  )
+  {
+    var rootObjectCollection = new Collection { name = RhinoDoc.ActiveDoc.Name ?? "Unnamed document" };
+    int count = 0;
+
+    Dictionary<int, Collection> layerCollectionCache = new();
+    // TODO: Handle blocks.
+    foreach (RhinoObject rhinoObject in rhinoObjects)
+    {
+      if (cts.IsCancellationRequested)
+      {
+        throw new OperationCanceledException(cts.Token);
+      }
+
+      // 1. get object layer
+      var layer = RhinoDoc.ActiveDoc.Layers[rhinoObject.Attributes.LayerIndex];
+
+      // 2. get or create a nested collection for it
+      var collectionHost = GetHostObjectCollection(layerCollectionCache, layer, rootObjectCollection);
+      var applicationId = rhinoObject.Id.ToString();
+
+      // 3. get from cache or convert:
+      // What we actually do here is check if the object has been previously converted AND has not changed.
+      // If that's the case, we insert in the host collection just its object reference which has been saved from the prior conversion.
+      /*Base converted;
+      if (
+        !modelCard.ChangedObjectIds.Contains(applicationId)
+        && _convertedObjectReferences.TryGetValue(applicationId + modelCard.ProjectId, out ObjectReference value)
+      )
+      {
+        converted = value;
+      }
+      else
+      {
+        converted = converter.ConvertToSpeckle(rhinoObject);
+        converted.applicationId = applicationId;
+      }*/
+
+      var converted = converter.ConvertToSpeckle(rhinoObject);
+      converted.applicationId = applicationId;
+
+      // 4. add to host
+      collectionHost.elements.Add(converted);
+      _basicConnectorBinding.Commands.SetModelProgress(
+        modelCard.ModelCardId,
+        new ModelCardProgress { Status = "Converting", Progress = (double)++count / rhinoObjects.Count }
+      );
+
+      // NOTE: useful for testing ui states, pls keep for now so we can easily uncomment
+      // Thread.Sleep(550);
+    }
+
+    // 5. profit
+    return rootObjectCollection;
+  }
+
+  /// <summary>
+  /// Returns the host collection based on the provided layer. If it's not found, it will be created and hosted within the the rootObjectCollection.
+  /// </summary>
+  /// <param name="layerCollectionCache"></param>
+  /// <param name="layer"></param>
+  /// <param name="rootObjectCollection"></param>
+  /// <returns></returns>
+  private Collection GetHostObjectCollection(
```

Comment:
We probably should look to inject this kind of things

Reply:
Yup, the entire send binding is already bigger than I feel comfortable with 🤣